### ISSUE TITLE
Ray can't be restarted after several hours of use

### ISSUE THREAD

**Issue by Kuurusch** [ray-project/ray#8080]
Hallo,

So I'm using ray 0.8.2 with python 3.6 on a Kubuntu 18.04LTS and my problem is the following:

I reinitialize for several tests ray with init() and at the end shuting it down with shutdown() and this works great for several hours of many tests. But after some time I've got the following messages, strangly in two trials the message were different:

First try:
```
2020-04-17 23:22:13,630	INFO resource_spec.py:212 -- Starting Ray with 82.62 GiB memory available for workers and up to 18.63 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-04-17 23:22:14,148	WARNING services.py:850 -- Redis failed to start, retrying now.
2020-04-17 23:22:14,828	INFO services.py:1078 -- View the Ray dashboard at localhost:8265
Could not connect to 127.0.0.1: 41937
Traceback (most recent call last):
  File "/home/sia1/pyCharm/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_comm.py", line 449, in start_client
    s.connect((host, port))
socket.timeout: timed out
Traceback (most recent call last):
  File "/home/sia1/pyCharm/plugins/python-ce/helpers/pydev/pydevd.py", line 2127, in <module>
    main()
  File "/home/sia1/pyCharm/plugins/python-ce/helpers/pydev/pydevd.py", line 2009, in main
    dispatcher.connect(host, port)
  File "/home/sia1/pyCharm/plugins/python-ce/helpers/pydev/pydevd.py", line 1784, in connect
    self.client = start_client(self.host, self.port)
  File "/home/sia1/pyCharm/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_comm.py", line 449, in start_client
    s.connect((host, port))
socket.timeout: timed out
```

Second try:

```
2020-04-18 04:03:18,889	INFO resource_spec.py:212 -- Starting Ray with 73.34 GiB memory available for workers and up to 18.63 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-04-18 04:03:19,330	INFO services.py:1078 -- View the Ray dashboard at localhost:8265
E0418 04:03:20.365157 52773 raylet_client.cc:60] Retrying to connect to socket for pathname /tmp/ray/session_2020-04-18_04-03-18_842152_52773/sockets/raylet (num_attempts = 1, num_retries = 5)
E0418 04:03:20.866624 52773 raylet_client.cc:60] Retrying to connect to socket for pathname /tmp/ray/session_2020-04-18_04-03-18_842152_52773/sockets/raylet (num_attempts = 2, num_retries = 5)
E0418 04:03:21.366780 52773 raylet_client.cc:60] Retrying to connect to socket for pathname /tmp/ray/session_2020-04-18_04-03-18_842152_52773/sockets/raylet (num_attempts = 3, num_retries = 5)
E0418 04:03:21.866927 52773 raylet_client.cc:60] Retrying to connect to socket for pathname /tmp/ray/session_2020-04-18_04-03-18_842152_52773/sockets/raylet (num_attempts = 4, num_retries = 5)
F0418 04:03:22.367069 52773 raylet_client.cc:69] Could not connect to socket /tmp/ray/session_2020-04-18_04-03-18_842152_52773/sockets/raylet
*** Check failure stack trace: ***
    @     0x7f97b6303efd  google::LogMessage::Fail()
    @     0x7f97b630536c  google::LogMessage::SendToLog()
    @     0x7f97b6303bd9  google::LogMessage::Flush()
    @     0x7f97b6303df1  google::LogMessage::~LogMessage()
    @     0x7f97b60b2cc9  ray::RayLog::~RayLog()
    @     0x7f97b5fdf3b1  ray::raylet::RayletConnection::RayletConnection()
    @     0x7f97b5fdfd21  ray::raylet::RayletClient::RayletClient()
    @     0x7f97b5f92eb3  ray::CoreWorker::CoreWorker()
    @     0x7f97b5f09f7b  __pyx_tp_new_3ray_7_raylet_CoreWorker()
    @           0x551b15  (unknown)
    @           0x5aa6ec  _PyObject_FastCallKeywords
    @           0x50abb3  (unknown)
    @           0x50c5b9  _PyEval_EvalFrameDefault
    @           0x508245  (unknown)
    @           0x50a080  (unknown)
    @           0x50aa7d  (unknown)
    @           0x50d390  _PyEval_EvalFrameDefault
    @           0x5065b4  (unknown)
    @           0x50aa7d  (unknown)
    @           0x50c5b9  _PyEval_EvalFrameDefault
    @           0x508245  (unknown)
    @           0x50a080  (unknown)
    @           0x50aa7d  (unknown)
    @           0x50c5b9  _PyEval_EvalFrameDefault
    @           0x509455  _PyFunction_FastCallDict
    @           0x595311  (unknown)
    @           0x54a6ff  (unknown)
    @           0x551b81  (unknown)
    @           0x5aa6ec  _PyObject_FastCallKeywords
    @           0x50abb3  (unknown)
    @           0x50c5b9  _PyEval_EvalFrameDefault
    @           0x508245  (unknown)
```

Do you have any idea, what could be here the problem?
In the meantime I've updated to ray 0.8.4 and it is running at the moment...
I've further introduced a sleep of 10 seconds after shutdown(), hoping this would help somehow.
I don't think it is a memory issue, because the memory-usage stays constant.

**Comment by Kuurusch** [comment]
Update:

So my third try is still running, but what I've already observed is by chance, that it had already the same problem like in the first try, but this time it manached somehow to reconnect. So I've got the following warning, but afterwards it run further without problems (till now, the csript is still running...):

```
2020-04-18 17:06:10,629	INFO resource_spec.py:212 -- Starting Ray with 61.13 GiB memory available for workers and up to 30.19 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-04-18 17:06:10,772	WARNING services.py:916 -- Redis failed to start, retrying now.
2020-04-18 17:06:11,230	INFO services.py:1148 -- View the Ray dashboard at localhost:8266
```

So it seems to have from time to time problems, connecting to Redis. Should I increase somewhere the timeout or the number of reconnection-attempts and how would I do this?

**Comment by Kuurusch** [comment]
Update 2:

Ok, now it worked! Don't know, if it was the Ray-Update, the introduced sleep(10) after shutdown() or just by chance! Will investigate it further and give you an update, if I have new information...

**Comment by robertnishihara** [comment]
@Kuurusch thanks for the update. Let us know :)

**Comment by janblumenkamp** [comment]
I also noticed a few times that Ray (using it with RLlib) just stopped working after a few hours with exactly one CPU core on 100% load. The process could then neither stopped with SIGINT nor with SIGQUIT. I was running in a tmux session and I could kill the session, but the memory would still be allocated and there were a few Ray processes left running. I restarted the machine and didn't investigate any further. Not sure if this is really related to this issue, but I will try to find out more if this happens again for me.

**Comment by Kuurusch** [comment]
Update2:
Ok, I've let it run a second time with the measures discribed above and it worked again. So it was much likely not by chance, that it worked. And then I've let it run a third time without the sleep(10) and it worked too. So I think the problem was fixed in ray 0.8.4! Thanks a lot!
A last question, would it be possible for me to set the timeout and the times of reconnection-tries to Redis, so that I could increase them?

**Comment by richardliaw** [comment]
@janblumenkamp can you open a new issue for that?
@Kuurusch re: timeout, can you file a question/feature request?

Just wanted to close this issue because the original issue is resolved (and single-issue threads is much easier on the project maintainers)
